src/intrinsic/simd.rs

```diff
@@ -432,30 +431,47 @@ pub fn generic_simd_intrinsic<'a, 'gcc, 'tcx>(bx: &mut Builder<'a, 'gcc, 'tcx>,
 
         let vec_ty = bx.cx.type_vector(elem_ty, in_len);
 
-        let (intr_name, fn_ty) =
+        let (intr_name, num_args) =
             match name {
-                sym::simd_ceil => ("ceil", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_fabs => ("fabs", bx.type_func(&[vec_ty], vec_ty)), // TODO(antoyo): pand with 170141183420855150465331762880109871103
-                sym::simd_fcos => ("cos", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_fexp2 => ("exp2", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_fexp => ("exp", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_flog10 => ("log10", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_flog2 => ("log2", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_flog => ("log", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_floor => ("floor", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_fma => ("fma", bx.type_func(&[vec_ty, vec_ty, vec_ty], vec_ty)),
-                sym::simd_fpowi => ("powi", bx.type_func(&[vec_ty, bx.type_i32()], vec_ty)),
-                sym::simd_fpow => ("pow", bx.type_func(&[vec_ty, vec_ty], vec_ty)),
-                sym::simd_fsin => ("sin", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_fsqrt => ("sqrt", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_round => ("round", bx.type_func(&[vec_ty], vec_ty)),
-                sym::simd_trunc => ("trunc", bx.type_func(&[vec_ty], vec_ty)),
+                sym::simd_ceil => ("ceil", 1),
+                sym::simd_fabs => ("fabs", 1), // TODO(antoyo): pand with 170141183420855150465331762880109871103
+                sym::simd_fcos => ("cos", 1),
+                sym::simd_fexp2 => ("exp2", 1),
+                sym::simd_fexp => ("exp", 1),
+                sym::simd_flog10 => ("log10", 1),
+                sym::simd_flog2 => ("log2", 1),
+                sym::simd_flog => ("log", 1),
+                sym::simd_floor => ("floor", 1),
+                sym::simd_fma => ("fma", 3),
+                sym::simd_fpowi => ("__builtin_powi", 2),
+                sym::simd_fpow => ("pow", 2),
+                sym::simd_fsin => ("sin", 1),
+                sym::simd_fsqrt => ("sqrt", 1),
+                sym::simd_round => ("round", 1),
+                sym::simd_trunc => ("trunc", 1),
                 _ => return_error!("unrecognized intrinsic `{}`", name),
             };
-        let llvm_name = &format!("llvm.{0}.v{1}{2}", intr_name, in_len, elem_ty_str);
-        let function = intrinsic::llvm::intrinsic(llvm_name, &bx.cx);
-        let function: RValue<'gcc> = unsafe { std::mem::transmute(function) };
-        let c = bx.call(fn_ty, function, &args.iter().map(|arg| arg.immediate()).collect::<Vec<_>>(), None);
+        let builtin_name = format!("{0}{1}", intr_name, elem_ty_str);
+        let funcs = bx.cx.functions.borrow();
+        let function = funcs.get(&builtin_name).unwrap();
```

Comment:
Please use `expect()` instead of `unwrap()`.
I'll add the clippy lint for that.

Reply:
I had `unwrap()` because this lookup should never fail; we declare all of these builtins [here](https://github.com/rust-lang/rustc_codegen_gcc/blob/master/src/context.rs#L173-L187), so unwrapping should be safe to do.  That said, this should probably be a little more expressive in what failed and why if the lookup fails.

Would something like `unwrap_or_else(|_| panic!("unable to find builtin function {}", builtin_name))` be acceptable?  If we failed to lookup a builtin, it would be nice to report what builtin we missed.

Reply:
Yes, that would be perfect.

Reply:
Fixed in 9691257.